AX: WebKit fails to provide accurate geometry to ATs for opacity:0 elements, causing various issues like not being able to press custom controls

twilco opened the pull request:
#### df1a17143317681864b4b550671886cda6ff5003
<pre>
AX: WebKit fails to provide accurate geometry to ATs for opacity:0 elements, causing various issues like not being able to press custom controls
<a href="https://bugs.webkit.org/show_bug.cgi?id=291879">https://bugs.webkit.org/show_bug.cgi?id=291879</a>
<a href="https://rdar.apple.com/149726630">rdar://149726630</a>

Reviewed by Joshua Hoffman.

It&apos;s common for web developers to create custom controls by hiding an input element with opacity:0, then use CSS to
place a nicer looking custom control on top that can interact with pointer devices. In these cases, developers rely on
the visually-hidden input element (or whatever element) for accessibility semantics. Currently, we return very
inaccurate geometry for these opacity:0 elements because accessibility relies on paint to get geometry, and opacity:0
things are not painted. This can have various bad side effects, e.g. we may produce a rect with zero width and height,
which can cause ATs to ignore the element outright, or produce a rect with a very wrong position, which is confusing
for low vision users and can cause scroll jumping.

Fix this by painting opacity:0 elements for accessibility paints.

* LayoutTests/accessibility/opacity-0-bounding-box-expected.txt: Added.
* LayoutTests/accessibility/opacity-0-bounding-box.html: Added.
* LayoutTests/platform/glib/TestExpectations: Disable new test.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/opacity-0-bounding-box-expected.txt: Added.
* LayoutTests/platform/mac-wk1/TestExpectations: Disable new test.
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper _accessibilityPageRelativeLocation]):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(attributeValueForTesting):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayerWithEffects):
* Tools/DumpRenderTree/AccessibilityUIElement.cpp:
(getPageXCallback):
(getPageYCallback):
(AccessibilityUIElement::getJSClass):
* Tools/DumpRenderTree/AccessibilityUIElement.h:
* Tools/DumpRenderTree/ios/AccessibilityUIElementIOS.mm:
(AccessibilityUIElement::pageX):
(AccessibilityUIElement::pageY):
* Tools/DumpRenderTree/mac/AccessibilityUIElementMac.mm:
(AccessibilityUIElement::pageX):
(AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityUIElementAtspi.cpp:
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::attributeValue):
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/playstation/AccessibilityUIElementPlayStation.cpp:
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/win/AccessibilityUIElementWin.cpp:
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):

Canonical link: <a href="https://commits.webkit.org/293993@main">https://commits.webkit.org/293993@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a7acca779d2f041cf499fead957f0fbbc76e22a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100458 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20110 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10409 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105595 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51046 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102499 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20417 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28584 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76496 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33547 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103465 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15648 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90746 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56852 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15465 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8749 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50422 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85381 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8828 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107949 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27576 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20242 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85449 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27939 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86946 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84987 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21629 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29672 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7405 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21541 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27511 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32761 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27322 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30640 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28880 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on LayoutTests/accessibility/opacity-0-bounding-box.html:
```diff
@@ -0,0 +1,65 @@
+    <input type="radio" style="width: 300px; height: 300px">
+    Opacity: 0 text
+    <img src="./resources/cake.png" alt="adorned white cake on a table">
+    <!-- <button>Press</button> -->
```
Can this commented out button be removed?

hoffmanjoshua commented on LayoutTests/accessibility/opacity-0-bounding-box.html:
```diff
@@ -0,0 +1,65 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN"><!-- webkit-test-runner [ runSingly=true -->
```
```html
<!-- webkit-test-runner [ runSingly=true -->
```
There's a missing a closing bracket here

hoffmanjoshua commented on LayoutTests/accessibility/opacity-0-bounding-box-expected.txt:
```diff
@@ -0,0 +1,29 @@
+
+PASS: fooText.role.toLowerCase().includes('statictext') === true
+Foo x: -9992
+Foo y: 10949
```
It looks like the bots/EWS produce different positions for all/most of these, likely due to using different size devices. I ran into this on a recent WIP patch as well. I wonder if we can verify positions in some way that are relative to something that isn't changing, rather than the device? What do you think?

hoffmanjoshua reviewed (approved):
Thanks for adding the new testing infrastructure! That’ll be super helpful moving forward

twilco committed:
AX: WebKit fails to provide accurate geometry to ATs for opacity:0 elements, causing various issues like not being able to press custom controls
https://bugs.webkit.org/show_bug.cgi?id=291879
rdar://149726630

Reviewed by Joshua Hoffman.

It's common for web developers to create custom controls by hiding an input element with opacity:0, then use CSS to
place a nicer looking custom control on top that can interact with pointer devices. In these cases, developers rely on
the visually-hidden input element (or whatever element) for accessibility semantics. Currently, we return very
inaccurate geometry for these opacity:0 elements because accessibility relies on paint to get geometry, and opacity:0
things are not painted. This can have various bad side effects, e.g. we may produce a rect with zero width and height,
which can cause ATs to ignore the element outright, or produce a rect with a very wrong position, which is confusing
for low vision users and can cause scroll jumping.

Fix this by painting opacity:0 elements for accessibility paints.

* LayoutTests/accessibility/opacity-0-bounding-box-expected.txt: Added.
* LayoutTests/accessibility/opacity-0-bounding-box.html: Added.
* LayoutTests/platform/glib/TestExpectations: Disable new test.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/opacity-0-bounding-box-expected.txt: Added.
* LayoutTests/platform/mac-wk1/TestExpectations: Disable new test.
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper _accessibilityPageRelativeLocation]):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(attributeValueForTesting):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayerWithEffects):
* Tools/DumpRenderTree/AccessibilityUIElement.cpp:
(getPageXCallback):
(getPageYCallback):
(AccessibilityUIElement::getJSClass):
* Tools/DumpRenderTree/AccessibilityUIElement.h:
* Tools/DumpRenderTree/ios/AccessibilityUIElementIOS.mm:
(AccessibilityUIElement::pageX):
(AccessibilityUIElement::pageY):
* Tools/DumpRenderTree/mac/AccessibilityUIElementMac.mm:
(AccessibilityUIElement::pageX):
(AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityUIElementAtspi.cpp:
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::attributeValue):
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/playstation/AccessibilityUIElementPlayStation.cpp:
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):
* Tools/WebKitTestRunner/InjectedBundle/win/AccessibilityUIElementWin.cpp:
(WTR::AccessibilityUIElement::pageX):
(WTR::AccessibilityUIElement::pageY):

Canonical link: https://commits.webkit.org/293993@main
